jobs/coin_dbc.go

```diff
@@ -0,0 +1,241 @@
+package jobs
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"api.audius.co/config"
+	"api.audius.co/database"
+	"api.audius.co/logging"
+	"api.audius.co/solana/spl/programs/meteora_dbc"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/jackc/pgx/v5"
+	"github.com/krazyTry/meteora-go/dbc"
+	"go.uber.org/zap"
+)
+
+type CoinDBCJob struct {
+	meteoraClient *meteora_dbc.Client
+	dbc           *dbc.DBC
+	pool          database.DbPool
+	logger        *zap.Logger
+
+	mutex     sync.Mutex
+	isRunning bool
+}
+
+func NewCoinDBCJob(config config.Config, pool database.DbPool) *CoinDBCJob {
+	logger := logging.NewZapLogger(config).Named("CoinDbcJob")
+	rpcClient := rpc.New(config.SolanaConfig.RpcProviders[0])
+	meteoraClient := meteora_dbc.NewClient(rpcClient, logger)
+	dbc := dbc.NewDBC(rpcClient)
+
+	return &CoinDBCJob{
+		meteoraClient: meteoraClient,
+		logger:        logger,
+		pool:          pool,
+		dbc:           dbc,
+	}
+}
+
+// ScheduleEvery runs the job every `duration` until the context is cancelled.
+func (j *CoinDBCJob) ScheduleEvery(ctx context.Context, duration time.Duration) *CoinDBCJob {
+	go func() {
+		ticker := time.NewTicker(duration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				j.logger.Info("Job started")
+				j.Run(ctx)
+			case <-ctx.Done():
+				j.logger.Info("Job shutting down")
+				return
+			}
+		}
+	}()
+	return j
+}
+
+// Run executes the job once
+func (j *CoinDBCJob) Run(ctx context.Context) {
+	if err := j.run(ctx); err != nil {
+		j.logger.Error("Job run failed", zap.Error(err))
+	} else {
+		j.logger.Info("Job completed successfully")
+	}
+}
+
+// For each artist coin in the database, fetches the latest stats from the Meteora DBC
+// updates the artist_coin_pools table. Ensures only one instance runs at a time.
+func (j *CoinDBCJob) run(ctx context.Context) error {
+	j.mutex.Lock()
+	if j.isRunning {
+		j.mutex.Unlock()
+		return fmt.Errorf("job is already running")
+	}
+	j.isRunning = true
+	j.mutex.Unlock()
+	defer func() {
+		j.mutex.Lock()
+		j.isRunning = false
+		j.mutex.Unlock()
+	}()
+
+	count, err := GetTokenCount(ctx, j.pool)
+	if err != nil {
+		return fmt.Errorf("error getting token count: %w", err)
+	}
+
+	for offset := 0; offset < count; offset += tokenPageSize {
+		batch, err := GetTokenBatch(ctx, j.pool, tokenPageSize, offset)
+		if err != nil {
+			return fmt.Errorf("error getting token batch: %w", err)
+		}
+
+		for _, coin := range batch {
+			baseMint := solana.MustPublicKeyFromBase58(coin.Mint)
+			pool, err := j.dbc.GetPoolByBaseMint(ctx, baseMint)
+			if err != nil {
+				j.logger.Error("pool lookup failed", zap.String("mint", coin.Mint), zap.Error(err))
+				continue
+			}
+			if pool == nil {
+				j.logger.Debug("No pool found for mint", zap.String("mint", coin.Mint))
```

Comment:
Is this an unexpected case? Just wondering if it should be surfaced in axiom logs.

Reply:
the above case handles error where for some reason we can't talk to solana